File: client/src/nix_base32.rs
// Nix uses a custom base32 alphabet
const NIX_BASE32_ALPHABET: &[u8] = b"0123456789abcdfghijklmnpqrsvwxyz";

pub fn encode(input: &[u8]) -> String {
    if input.is_empty() {
        return String::new();
    }

    // Calculate the output length
    let len = (input.len() * 8 - 1) / 5 + 1;

    let mut result = String::with_capacity(len);

    // Process from the highest bit position down to 0
    for n in (0..len).rev() {
        let b = n * 5;
        let i = b / 8;
        let j = b % 8;

        // Extract 5 bits starting at bit position b
        let mut c = if i < input.len() { input[i] >> j } else { 0 };

        if i + 1 < input.len() && j > 3 {
            c |= input[i + 1] << (8 - j);
        }

        result.push(NIX_BASE32_ALPHABET[(c & 0x1f) as usize] as char);
    }

    result
}

pub fn hash_to_nix_string(algo: &str, hash: &[u8]) -> String {
    format!("{}:{}", algo, encode(hash))
}

#[cfg(test)]
mod tests {
    use super::*;
    use sha2::{Digest, Sha256};

    #[test]
    fn test_nix_base32_encode() {
        // Test vectors from Nix
        assert_eq!(encode(b""), "");
        assert_eq!(encode(b"\x00"), "00");
        assert_eq!(encode(b"\xff"), "7z");

        // Test a known hash
        let mut hasher = Sha256::new();
        hasher.update(b"hello");
        let hash = hasher.finalize();
        let encoded = encode(&hash);

        // This is the expected nix32 encoding of SHA256("hello")
        // Verified with: echo -n "hello" | nix hash file --type sha256 --base32 /dev/stdin
        assert_eq!(
            encoded,
            "094qif9n4cq4fdg459qzbhg1c6wywawwaaivx0k0x8xhbyx4vwic"
        );
    }

    #[test]
    fn test_hash_to_nix_string() {
        let hash = [0x12, 0x34, 0x56, 0x78];
        let result = hash_to_nix_string("sha256", &hash);
        assert!(result.starts_with("sha256:"));
    }
}

File: client/nixbase32.go
package client

import (
	"fmt"
	"strings"
)

// nixbase32 implements Nix's base32 encoding scheme.
// This is a custom base32 encoding that omits certain characters (e, o, u, t)
// to avoid confusion and uses a specific alphabet.

const nixBase32Alphabet = "0123456789abcdfghijklmnpqrsvwxyz"

// EncodeNixBase32 encodes bytes into Nix's base32 format.
// This implementation is based on Nix's BaseNix32::encode in src/libutil/base-nix-32.cc.
func EncodeNixBase32(input []byte) string {
	if len(input) == 0 {
		return ""
	}

	// Calculate the encoded length: (len * 8 - 1) / 5 + 1
	length := (len(input)*8-1)/5 + 1

	result := make([]byte, 0, length)

	// The C++ implementation builds the string by iterating n from high to low
	// and appending characters. This means the character for n=length-1 comes first.
	for n := length - 1; n >= 0; n-- {
		b := n * 5
		i := b / 8
		j := b % 8

		// Extract 5 bits starting at position (i, j)
		var c byte
		if i < len(input) {
			c = input[i] >> j
		}

		if i+1 < len(input) {
			c |= input[i+1] << (8 - j)
		}

		// Take only the lower 5 bits and map to alphabet
		result = append(result, nixBase32Alphabet[c&0x1f])
	}

	return string(result)
}

func nixBase32Len(n int) int {
	if n == 0 {
		return 0
	}

	return (n*8-1)/5 + 1
}

// DecodeNixBase32 is the inverse of EncodeNixBase32 (libutil/base-nix-32.cc).
func DecodeNixBase32(s string) ([]byte, error) {
	n := len(s) * 5 / 8
	if nixBase32Len(n) != len(s) {
		return nil, fmt.Errorf("invalid nix32 length %d", len(s))
	}

	out := make([]byte, n)

	for pos := range len(s) {
		c := s[len(s)-1-pos]

		idx := strings.IndexByte(nixBase32Alphabet, c)
		if idx < 0 {
			return nil, fmt.Errorf("invalid nix32 character %q", c)
		}

		digit := byte(idx) //nolint:gosec // idx < 32

		b := pos * 5
		i := b / 8
		j := b % 8
		out[i] |= digit << j

		carry := digit >> (8 - j)
		if i+1 < n {
			out[i+1] |= carry
		} else if carry != 0 {
			return nil, fmt.Errorf("invalid nix32 string %q: trailing bits", s)
		}
	}

	return out, nil
}
